### DISCUSSION TITLE
JetBrains & VS Code are using all the free storage quota

### DISCUSSION THREAD

**Discussion by huw** [community/community#57767] [Codespaces] [answered]
I started getting storage notifications a few months ago, when I previously hadn’t. I know I don’t have a huge repository, so I decided to use `ncdu` to check it out. I found that:

- `/workspaces/.codespaces/shared/editors/jetbrains` is using 11.2 GB to store a number of JetBrains editors
- `/.codespaces` and `/vscode` use 5.2 GB each. I’m not sure VS Code actually takes that much to store, but at the very least nothing is running from `/.codespaces/bin/cache`, whereas the VS Code process definitely runs from `/vscode`.

Deleting these doesn’t break anything for me, because I use VS Code for Web exclusively. However, these are re-added every time the Codespace launches. I wasn’t able to fix it by using a `postStartCommand` either, but I didn’t try very hard because I don’t think it’ll actually remove them from the billable usage.

Now, I don’t know if these files are actually counted toward the storage quota (I hope they aren’t, but I can’t think of anything else, and their size lines up with the quota I’m consuming each month). I would hope, to be honest, that keeping a single Codespace in storage doesn’t automatically consume the free quota; this seems against the spirit of having a free quota in the first place. Can someone from GitHub explain what’s going on?

**Comment by cwndrws** [comment, marked as answer]
Hi 👋 thanks for writing in. The intention of our storage billing is that you are only billed for the storage that you use, which in the case of using the default devcontainer, is just things you've added to the image (your repo, any system configuration and any tools you've manually installed). I don't believe the directories you mention, `/workspaces/.codespaces/shared/`, `/.codespaces/` or `/vscode/` are counted toward your storage usage. From [our docs about how we bill storage](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces#about-billing-for-storage-usage):

> This includes any files you use in a codespace, such as cloned repositories, configuration files, data loaded to the codespace (for example as input or output of the software running in the repository), and extensions, among others.

and

> If you use the default image, the storage usage for your codespace will be based on the files in your repository, and any files you subsequently add to the codespace. If you use an alternative base image, then the resulting container and all of the files in the codespace will be counted as used storage.

If you're looking into how to understand what is getting counted toward the storage you're being billed for, how big is the repo that you're using in codespaces? Also are you using prebuilds? We have had some users not realize that you get billed for prebuild storage in addition to the storage used by your running codespaces.

**Reply by huw** [reply]
Ah! I’m not using the default image because I needed to upgrade to Debian 12. Can you explain in more detail what counts as the ‘default’ image—does using any Dockerfile mean I’m non-default? What if I base my Dockerfile off a mcr.microsoft.com image to customise things slightly?

**Reply by cwndrws** [reply]
I think in that case, the whole container image would be counted toward your storage usage since I'm assuming you're using a different image entirely. The exception for the default image takes the percentage of matching docker layers into account, so if your container is based on the default one (`mcr.microsoft.com/devcontainers/universal`) and a high percentage of the layers match the default (this would be the case for "slight" customizations like you mention), the image shouldn't be counted against your storage usage.

**Reply by huw** [reply]
Fantastic, I’ve rebased my image and will check back in a few days to see if it’s fixed. I’m not sure where the Codespaces team takes feedback, but it would be nice for custom images to exclude those folders from the storage quota, or otherwise have a setting to choose what gets installed ^_^ — cheers again

**Reply by huw** [reply]
Just to give an update here—I can’t use the `universal` image because it’s on an outdated version of Linux, so I’m still forced to deal with the huge file sizes. However, I was able to successfully add `"postStartCommand": "rm -rf /workspaces/.codespaces/shared/editors/jetbrains"` to my `devcontainer.json`, and I’ll see if this helps with the storage quota over time. I can’t attack the VS Code cache because I’m running non-root, but JetBrains was the bigger killer anyway.

**Reply by cwndrws** [reply]
Like I said in my initial answer, I don't believe `/workspaces/.codespaces/shared/` is counted toward your storage usage, but report back if that seems to not be the case.

**Reply by huw** [reply]
Here’s my storage report. It’s a little hard to disentangle because on the date with the red line, I deleted my old Codespace and created a new one _as well as_ adding that command. My storage use dropped precipitously.

![IMG_0080](https://github.com/community/community/assets/2456132/cf4af9e7-baf6-40e5-808a-8ac1d2327715)

Regardless, if recreating my Codespace dropped the storage by that much, it’s definitely counting something on the OS that is added after my Docker setup, and `ncdu` is pointing toward that thing being the shared editors. Happy to send more information as necessary to investigate this further, but at least I have a good workaround for now ^_^

**Reply by cwndrws** [reply]
Thanks! I'll report this to the folks that handle storage billing for clarification.

**Reply by wolfmanstout** [reply]
Any updates on this? I'm also not using the default/universal base image, and I was surprised to see myself run out of free quota. From all the df's and du's I've run, I'm pretty confident this is the problem, as it is responsible for 12G of usage. I don't use JetBrains, and would be happy to have an option to simply not add all these jetbrains files. When I run as a local dev container (with the same devcontainer.json) none of these files are there. I guess I can add this postCreateCommand, but it's polluting a config file that's not only intended for codespaces with a fix for a very codespaces-specific issue that JetBrains users wouldn't want.

**Reply by stylpe** [reply]
Me too! Same path, same size. I have a .devcontainer specifying a custom image, which is probably part of the repro for this since as you say this will cause all files in the codespace to be counted. But I haven't done anything to ask for four JetBrains products to be installed in my codespace 😅

I've deleted the folder now and will check in a couple of days if usage shrinks.

I'm guessing this could be something that's done by default to have an ideal out-of-the-box experience for JetBrains users? But this side effect is nasty.

**Reply by lepinsk** [reply]
I'm running into the same issue; Codespaces reporting high storage usage when my repo is quite tiny (and I'm using the default image).

It seems to be down to files being stored in `/workspaces/.codespaces/shared/editors/jetbrains`:

```bash
@me ➜ .../.codespaces/shared/editors/jetbrains $ du -hs ./*
3.4G    ./JetBrainsRider-2023.1
2.1G    ./RubyMine-2023.1.1
3.6G    ./idea-IU-231.8770.65
2.3G    ./pycharm-2023.1.1
```

**Reply by AG-DavidG** [reply]
I'm running into the exact same problem. 
Keep getting warnings that my space is running low, and there is not enough space to run workflows that I need.
I deleted that directory and can only hope that it doesn't come back.

**Reply by huw** [reply]
@AG-DavidG That won’t work, it rebuilds very regularly (possibly on boot?)

The only reliable way to solve this is to start with the universal base image and modify as necessary.

**Reply by ryanwinter** [reply]
Just ran into this issue, so are you saying to remove the ~/.codespaces/shared/editors/jetbrains in the Dockerfile with the rest of the setup?

**Reply by huw** [reply]
@ryanwinter No, the only solution is to source your Dockerfile from `mcr.microsoft.com/devcontainers/universal`, and then run whatever commands you need to run to get it to your ideal setup. This is the only image which doesn't count `~/.codespaces` toward the storage quota (I was unable to reliably delete the offending editors from other images in a way that didn't count toward my storage quota). If you don't want to use Debian 13, then you just have to accept the extra cost.

**Reply by ryanwinter** [reply]
Oh nice, this is the one I'm already using and yeah I confirmed that directory is mounted from elsewhere so not included. 

I still wish they had more granularity in they hardware options, or just bump the disk space up a little bit for all of them.

**Reply by Samigirma1** [reply]
I'm running into the same issue. My problem is less with billing and more about actually available capacity. My workflow requires building containers within the dev container. Container builds are consistently failing because I'm running out of disk space.

**Reply by YoungjuneKwon** [reply]
> @AG-DavidG That won’t work, it rebuilds very regularly (possibly on boot?)
> 
> The only reliable way to solve this is to start with the universal base image and modify as necessary.

I have always used the default image to work, but there have been same folders in all of my codespace.
Therefore, I think this problem might be occurred in any images.

Does anyone have other solutions for this problem?

**Comment by JamesHutchison** [comment]
I am seeing this issue and it is resulting in an error message about low disk space. I do not use JetBrains.

```
.../.codespaces/shared/editors/jetbrains $ du -h --max-depth=1
2.1G    ./RubyMine-2023.1.1
3.6G    ./idea-IU-231.8770.65
2.3G    ./pycharm-2023.1.1
3.4G    ./JetBrainsRider-2023.1
12G     .
```

```
df
Filesystem     1K-blocks     Used Available Use% Mounted on
overlay         32847680 29453500   1700096  95% /
tmpfs              65536        0     65536   0% /dev
shm                65536        0     65536   0% /dev/shm
/dev/loop4      32847680 29453500   1700096  95% /workspaces
/dev/sda1      123266624       96 116958792   1% /tmp
/dev/root       30298176 23791792   6490000  79% /vscode
```

Please do not write off this issue. Regardless of whether its billed or not, it is robbing disk space quota and creating an obnoxious error.

edit:
after deleting, df does not show recovered space
